<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2
import sqlalchemy as sa
from sqlalchemy import create_engine
import requests
from bs4 import BeautifulSoup

In [2]:
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    host='84.201.134.129',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    port='5432'
)

engine = create_engine('postgresql+psycopg2://skillfactory:cCkxxLVrDE8EbvjueeMedPKt@84.201.134.129:5432/project_sql')

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [3]:
query_3_1 = f'''
    select count(id)
    from vacancies
'''
df = pd.read_sql(sa.text(query_3_1), engine)

print(f"Количество вакансий в базе: {df['count'].iloc[0]}")

Количество вакансий в базе: 49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [4]:
query_3_2 = f'''
    select count(id)
    from employers
'''
df = pd.read_sql(sa.text(query_3_2), engine)

print(f"Количество работодателей в базе: {df['count'].iloc[0]}")

Количество работодателей в базе: 23501


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [5]:
query_3_3 = f'''
    select count(id)
    from areas
'''
df = pd.read_sql(sa.text(query_3_3), engine)

print(f"Количество регионов: {df['count'].iloc[0]}")

Количество регионов: 1362


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [6]:
query_3_4 = f'''
    select count(id)
    from industries
'''
df = pd.read_sql(sa.text(query_3_4), engine)

print(f"Количество сфер деятельности: {df['count'].iloc[0]}")

Количество сфер деятельности: 294


***

### В базе содержится информация о 49197 вакансиях в 294 сферах деятельности. Соотношение количества работодателей к количеству вакансий близко к 1:2. Воспользоваться базой для поиска работы/кандидатов на должность могут предствавители более 1300 городов.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [7]:
query_4_1 = f'''
    select 
        a.name as area, 
        count(v.id) as cnt
    from vacancies v
    join areas a on a.id = v.area_id
    group by a.name
    order by cnt desc
    limit 5
'''
df = pd.read_sql(sa.text(query_4_1), engine)

print("Города-лидеры по количеству вакансий:")
display(df)

Города-лидеры по количеству вакансий:


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [8]:
query_4_2 = f'''
    select count(*)
    from vacancies v
    where salary_from is not null or salary_to is not null
'''
df = pd.read_sql(sa.text(query_4_2), engine)

print(f"Количество вакансий с указанной з/п: {df['count'].iloc[0]}")

Количество вакансий с указанной з/п: 24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [9]:
query_4_3 = f'''
    select avg(salary_from) avg_from, avg(salary_to) avg_to
    from vacancies v
'''
df = pd.read_sql(sa.text(query_4_3), engine)

print(f"Среднее для нижней границы з/п: {df['avg_from'].iloc[0].round()}")
print(f"Среднее для верхней границы з/п: {df['avg_to'].iloc[0].round()}")

Среднее для нижней границы з/п: 71065.0
Среднее для верхней границы з/п: 110537.0


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [10]:
query_4_4 = f'''
    select schedule, employment, count(id)
    from vacancies 
    group by schedule, employment
    order by count(id) desc
'''
df = pd.read_sql(sa.text(query_4_4), engine)

print("Количество вакансий для сочетаний типов рабочего графика и типов трудоустройства:")
display(df)

Количество вакансий для сочетаний типов рабочего графика и типов трудоустройства:


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [33]:
query_4_5 = f'''
with y as (
select
    experience, 
    count(id) cnt
from vacancies 
group by experience
order by count(id)
),
p as (
select experience, round(avg(salary)::numeric, 2) avg_sal
from (
    select *,
    case
        when (salary_from is not null and salary_to is not null) then (salary_from + salary_to) / 2
        else coalesce(salary_from, 0) + coalesce(salary_to, 0) end as salary
    from vacancies
    where salary_from is not null or salary_to is not null
    ) as b
group by experience
)
select y.experience, cnt, avg_sal
from y
join p on y.experience = p.experience
order by 2
'''
df = pd.read_sql(sa.text(query_4_5), engine)

print("Количество вакансий в зависимости от требуемого опыта работы:")
display(df)

Количество вакансий в зависимости от требуемого опыта работы:


experience    cnt    avg_sal
0         Более 6 лет   1337  166256.08
1           Нет опыта   7197   40471.95
2       От 3 до 6 лет  14511  133210.92
3  От 1 года до 3 лет  26152   76540.82

***

* Города-лидеры по количеству вакансий предсказуемо - столицы (в настоящем или прошлом) стран таможенного союза
* Из всех представленных вакансий только в половине указана информация о з/п. Для ее прогнозирования для другой половины необходим анализ выборки с указанными значениями в привязке к признаками Опыт работы, Регион, Сфера деятельности, тип трудоустройства.
* Усредненный диапазон указанных з/п для выборки составляет 71000-110000. Цифры несут мало полезной информации без привязки к опыту работы, сфере деятельности, региону. 
* Наиболее популярным сочетанием типа рабочего графика с типом трудоустройства ожидаемо является "Полный день/полная занятость" 
* Более половины вакансий относится к кандидатам с опытом работы 1-3 года. С точки зрения работодателя это оправданный подход, т.к. позволяет найти  кандидатов, способных со старта выполнять обязанности, при этом претендующие на меньший, по сравнению с опытными кандидатами, уровень з/п: средняя з/п у кандидатов с опытом > 6 лет превышает зарплату кандидатов с опытом 1-3 года более чем в два раза. 

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [12]:
query_5_1 = f'''
select *
from
    (select 
        e.name, 
        count(v.id) cnt,
        row_number() over (order by count(v.id) desc) as rnk
    from vacancies v
    join employers e on e.id = v.employer_id
    group by e.name
    order by cnt desc) as r
where rnk = 1 or rnk =5
'''

df = pd.read_sql(sa.text(query_5_1), engine)

print("1-е, 5-е места (rnk) рейтинга работодателей по количеству вакансий:")
display(df)

1-е, 5-е места (rnk) рейтинга работодателей по количеству вакансий:


name   cnt  rnk
0         Яндекс  1933    1
1  Газпром нефть   331    5

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [13]:
query_5_2 = f'''
with x as (
select 
    a.id, 
    a.name,
    count (distinct v.employer_id) cnt_employers, 
    count(v.id) cnt_vacancies
from areas a
left join vacancies v on a.id = v.area_id
group by a.id, a.name
having count(v.id) = 0
)
select 
    x.name "Area w. no vacancy", 
    count(e.name) qty
from x
join employers e on e.area = x.id
group by x.name
order by 2 desc
limit 3
'''

df = pd.read_sql(sa.text(query_5_2), engine)

print("Регионы без вакансий - лидеры по количеству работодателей:")
display(df)

Регионы без вакансий - лидеры по количеству работодателей:


Area w. no vacancy  qty
0              Россия  410
1           Казахстан  207
2  Московская область   75

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [14]:
query_5_3 = f'''
select 
    e.name, 
    count(distinct v.area_id) cnt
from vacancies v
join employers e on e.id=v.employer_id
group by e.name
order by 2 desc
limit 10
'''

df = pd.read_sql(sa.text(query_5_3), engine)

print("Работодатели с самым широким охватом по количеству регионов, в которых они публикуют вакансии:")
display(df)

Работодатели с самым широким охватом по количеству регионов, в которых они публикуют вакансии:


name  cnt
0                      Яндекс  181
1                  Ростелеком  152
2                  Спецремонт  116
3      Поляков Денис Иванович   88
4                   ООО ЕФИН    71
5                  Совкомбанк   63
6                         МТС   55
7  ЭФКО, Управляющая компания   49
8                Почта России   48
9                        КРОН   48

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [15]:
query_5_4 = f'''
    with k as (
    select *
    from employers e
    left join employers_industries ei on e.id=ei.employer_id
    order by industry_id nulls first
    )
    select count(*)
    from k 
    where industry_id is null
'''

df = pd.read_sql(sa.text(query_5_4), engine)
print(f"Количество работодателей, не указавших сферу деятельности: {df['count'].iloc[0]}")

Количество работодателей, не указавших сферу деятельности: 8419


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [16]:
query_5_5 = f'''
select 
    e.name companie,  
    count(ei.industry_id), 
    row_number() over (order by e.name) as rnk
from employers_industries ei
join  employers e on e.id=ei.employer_id
group by e.name
having count(ei.industry_id) = 4
order by 1
limit 3
'''

df = pd.read_sql(sa.text(query_5_5), engine)
print(f"Компания с третьего места в списке работодателей (rnk #3), указавших 4 сферы деятельности: {df.loc[2,'companie']}")
display(df)

Компания с третьего места в списке работодателей (rnk #3), указавших 4 сферы деятельности: 2ГИС


companie  count  rnk
0  101 Интернет      4    1
1      21vek.by      4    2
2          2ГИС      4    3

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [17]:
query_5_6 = f'''
select count(*)
from
    (select *
    from employers_industries ei
    join industries i on i.id = ei.industry_id 
    where name = 'Разработка программного обеспечения') as j
'''

df = pd.read_sql(sa.text(query_5_6), engine)
print(f"Количество команий, занимающихся разработкой ПО: {df['count'].iloc[0]}")

Количество команий, занимающихся разработкой ПО: 3553


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [18]:
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' # Задаём адрес ресурса
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')
page_table = page.find("table", {'class': 'standard sortable'})
links = page_table.find_all('a')
mil_lst = [link.text for link in links[4:]]

mil_cities = tuple(mil_lst)

query_5_7 = f'''
with mil_city as (
select *
from (select a.name region, count(v.id) cnt
    from vacancies v
    join employers e on e.id = v.employer_id
    join areas a on a.id = v.area_id
    where e.name = 'Яндекс' 
    group by v.area_id, a.name) as d
where region in {mil_cities}
)
select *
from mil_city
union
(select 'Total', sum(mil_city.cnt)
from mil_city
)
order by cnt
'''

df = pd.read_sql(sa.text(query_5_7), engine)
display(df)

region    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7               Уфа   26.0
8            Самара   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            Total  485.0

***

* В среднем работодатели указывают 2 сферы деятельности, максимум: 16. 35% работодателей не указали сферу деятельности. Возможно агрегатору есть смысл сделать это поле обязательным для заполнения. Около 15% работодателей занимаются разработкой ПО, что подтверждает востребованность специалистов этой сферы.
* Лидерами по количеству вакансий являются крупные компании с распределенной сетью офисов: Яндекс, Ростелеком, Тинькофф, Сбер, Газпром нефть. Абсолютный лидер - Яндекс, на долю которого приходится около 4% всех представленнных вакансий; вакансии этой компании представлены во всех городах-милионниках РФ.
* Среди регионов без вакансий наибольшее количество работодателей наблюдается в крупных географических структурах уровня страны, области. Что может показаться достаточно странным и, скорее всего, говорит о нецелесообразности использования такого уровня территорий для описания вакансий: указание конкретного города более информативно

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [19]:
query_6_1 = f'''
    select count(name)
    from vacancies
    where lower(name) like '%data%' or lower(name) like '%данн%'
'''
df = pd.read_sql(sa.text(query_6_1), engine)
display(df['count'].iloc[0])

1771

In [20]:
# результат запроса
# 1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [21]:
query_6_2 = f'''
    select count(*)
    from vacancies
    where 
        (
        lower(name) like '%data%scientist%' or
        lower(name) like '%data%science%' or
        lower(name) like '%исследователь%данных%' or
        (name like '%ML%' and name not like '%HTML%') or
        lower(name) like '%machine%learning%' or
        lower(name) like '%машинн%обучен%'
        ) and
        (
        lower(name) like '%junior%' or
        experience = 'Нет опыта'or
        employment = 'Стажировка'
        )
'''
df = pd.read_sql(sa.text(query_6_2), engine)
display(df['count'].iloc[0])


51

In [22]:
# результат запроса
# 51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [23]:
query_6_3 = f'''
with ds as (
    select *
    from vacancies
    where 
        lower(name) like '%data%scientist%' or
        lower(name) like '%data%science%' or
        lower(name) like '%исследователь%данных%' or
        (name like '%ML%' and name not like '%HTML%') or
        lower(name) like '%machine%learning%' or
        lower(name) like '%машинн%обучен%'
)
select count(*)
from ds
where lower(key_skills) like '%sql%' or lower(key_skills) like '%postgres%'
'''
df = pd.read_sql(sa.text(query_6_3), engine)
display(df['count'].iloc[0])

201

In [24]:
# результат запроса
# 201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [25]:
query_6_4 = f'''
with ds as (
    select *
    from vacancies
    where 
        lower(name) like '%data%scientist%' or
        lower(name) like '%data%science%' or
        lower(name) like '%исследователь%данных%' or
        (name like '%ML%' and name not like '%HTML%') or
        lower(name) like '%machine%learning%' or
        lower(name) like '%машинн%обучен%'
)
select count(*)
from ds
where lower(key_skills) like '%python%'
'''
df = pd.read_sql(sa.text(query_6_4), engine)
display(df['count'].iloc[0])

351

In [26]:
# результат запроса
# 351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [27]:
query_6_5 = f'''
with ds as (
    select *
    from vacancies
    where 
        (
        lower(name) like '%data%scientist%' or
        lower(name) like '%data%science%' or
        lower(name) like '%исследователь%данных%' or
        (name like '%ML%' and name not like '%HTML%') or
        lower(name) like '%machine%learning%' or
        lower(name) like '%машинн%обучен%') 
)
select round(avg(skill_cnt)::numeric, 2) as avg
from    
    (select (length(key_skills) +1) - length(regexp_replace(key_skills, '\t', '', 'g')) as skill_cnt, 
    * 
    from ds) as x
'''
df = pd.read_sql(sa.text(query_6_5), engine)
display(df['avg'].iloc[0])

6.41

In [28]:
# результат запроса
# 6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [29]:
query_6_6 = f'''
with ds as (
    select *
    from vacancies
    where 
        (
        lower(name) like '%data%scientist%' or
        lower(name) like '%data%science%' or
        lower(name) like '%исследователь%данных%' or
        (name like '%ML%' and name not like '%HTML%') or
        lower(name) like '%machine%learning%' or
        lower(name) like '%машинн%обучен%') 
),
avg_salary as (
select *,
    case
        when (salary_from is not null and salary_to is not null) then (salary_from + salary_to) / 2
        else coalesce(salary_from, 0) + coalesce(salary_to, 0) end as salary
from ds
where salary_from is not null or salary_to is not null
)
select experience, round(avg(salary)::numeric, 0) avg_sal
from avg_salary
group by experience
'''
df = pd.read_sql(sa.text(query_6_6), engine)
display(df)



experience   avg_sal
0       От 3 до 6 лет  243115.0
1  От 1 года до 3 лет  139675.0
2           Нет опыта   74643.0

In [30]:
# результат запроса
# 243115.0

***

In [31]:
# выводы по предметному анализу

# Общий вывод по проекту

In [32]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования